api/classes/publisher.js

```diff
@@ -79,72 +90,151 @@ function deleteFile(file, root) {
+    // Then, delete them from the DB
+    var deletePublishedProject = publishedProject.destroy();
+    var deletePublishedFiles = publishedFiles.mapThen(function(file) {
+      return file.destroy();
+    });
+
+    return Promise.join(deletePublishedFiles, deletePublishedProject, function() {
+      // We deleted things!
+    });
+  });
 }
 
-exports.publish = function(project) {
+exports.publish = function publish(project) {
+  var user;
+
   return getUser(project)
-    .then(function(user) {
-      return modifyFiles(uploadFile, project, user);
-    })
-    .then(function(url) {
-      return project.set({
-        publish_url: url
-      }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully published ' +
-        project.get('title') +
-        ' to ' + project.get('publish_url')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
+  .then(function(u) {
+    user = u;
+  })
+  .then(function() {
+    return getPublishedProject(project)
+    .then(function(oldPublishedProject) {
+      if (oldPublishedProject) {
+        return clearPublishedProject(oldPublishedProject, project, user);
+      }
+    });
+  })
+  .then(function() {
+    return PublishedProject.forge({
+      project_id: project.get('id'),
+      title: project.get('title'),
+      tags: project.get('tags'),
+      description: project.get('description')
+    }).save();
+  })
+  .then(function(publishedProject) {
+    return File.query({
+      where: {
+        project_id: project.get('id')
+      }
+    }).fetchAll()
+    .then(function(files) {
+      if (files.length === 0) {
+        throw 'No files to be published!';
+      }
 
-      return Promise.reject(e);
+      return files.mapThen(function(file) {
+        return PublishedFiles.forge({
+          published_id: publishedProject.get('id'),
+          path: file.get('path'),
+          buffer: file.get('buffer')
+        }).save();
+      })
+      .then(function() {
+        return getPublishedFiles(publishedProject);
+      });
     });
+  })
+  .then(function(filesToBePublished) {
+    return modifyPublishedFiles(uploadFile, project, user, filesToBePublished);
+  })
+  .then(function(url) {
+    return project.set({
+      publish_url: url
+    }).save();
+  })
+  .then(function() {
+    log.info(
+      'Successfully published ' +
+      project.get('title') +
+      ' to ' + project.get('publish_url')
+    );
+  })
+  .catch(reject);
 };
 
-exports.unpublish = function(project) {
+exports.unpublish = function unpublish(project) {debugger;
   return getUser(project)
-    .then(function(user) {
-      return modifyFiles(deleteFile, project, user);
-    })
-    .then(function() {
-      return project.set({ publish_url: null }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully unpublished ' +
-        project.get('title')
-      );
+  .then(function(user) {
+    return getPublishedProject(project)
+    .then(function(publishedProject) {
```

Comment:
So lets do one or the other, either cache the user and keep the promise structure flat like you do in publish or follow this structure in publish too